src/login/logind-button.c

```diff
@@ -413,10 +520,42 @@ static int button_suitable(int fd) {
         return false;
 }
 
+static int button_get_modifier_mask(Button *b) {
+        int r;
+        unsigned long keys[KEY_MAX/ULONG_BITS+1];
+
+        r = ioctl(b->fd, EVIOCGKEY(sizeof(keys)), keys);
+        if (r < 0)
+                return -errno;
+
+        if (bitset_get(keys, KEY_LEFTSHIFT))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_LEFT_SHIFT;
+        if (bitset_get(keys, KEY_RIGHTSHIFT))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_RIGHT_SHIFT;
+
+        if (bitset_get(keys, KEY_LEFTCTRL))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_LEFT_CTRL;
+
+        if (bitset_get(keys, KEY_RIGHTCTRL))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_RIGHT_CTRL;
+
+        if (bitset_get(keys, KEY_LEFTMETA))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_LEFT_META;
+        if (bitset_get(keys, KEY_RIGHTCTRL))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_RIGHT_META;
+
+        if (bitset_get(keys, KEY_LEFTALT))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_LEFT_ALT;
+        if (bitset_get(keys, KEY_RIGHTALT))
+                b->button_modifier_mask &= ~BUTTON_MODIFIER_RIGHT_ALT;
+
```

Comment:
please use the keycode_modifier_table[] table info for this series of if checks. i.e. iterate through the array, look for the bits, and apply the mask

Reply:
maybe explicitly set b->button_modifier_mask to zero first